File: MAJORITY ELEMENT.py
'''
Majority Element 
Given an array A of N elements. Find the majority element in the array. A majority element in an array A of size N is an 
element that appears more than N/2 times in the array.

Example 1:

Input:
N = 3 
A[] = {1,2,3} 
Output: -1

Explanation: Since, each element in 
{1,2,3} appears only once so there 
is no majority element.

'''


class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        n = len(nums)
        if n==1:
            return nums[0]
        
        dict = {}
        candidate = -1
        
        for i in nums:
            if i not in dict:
                dict[i] = 1
            else:
                dict[i]+=1
                if dict[i]>n//2:
                    candidate = i
                    break
                    
        return candidate

    # Moore’s Voting Algorithm
    def majorityElement(self, nums: List[int]) -> int:
        
        candidate = -1
        counter = 0
        
        for j in range(len(nums)):
            if counter==0:
                candidate = nums[j]  
                counter = 1 #reset counter 
            elif nums[j]==candidate:
                counter+=1
            else:
                counter-=1
                
                
        return candidate

File: HASHMAP AND HEAP/MAJORITY ELEMENT.java

/**
 * 
Majority Element
Given an array nums of size n, return the majority element.
The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority 
element always exists in the array.

Example 1:

Input: nums = [3,2,3]
Output: 3
Example 2:

Input: nums = [2,2,1,1,1,2,2]
Output: 2
 */

import java.util.HashMap;
import java.util.Map;

class Solution {

    public int majorityElement(int[] nums){
        HashMap<Integer,Integer> map = new HashMap<>();

        for(int e : nums){
            map.put(ele,map.getOrDefault(ele,0)+1);
        }

        for(Map.Entry<Integer,Integer> entry : map.entrySet()){
            if(entry.getValue()>nums.length/2){
                return entry.getKey();
            }
        }
        return -1;
    }
    // Moore’s Voting Algorithm
    public int majorityElement2(int[] nums) {
        //find candidate
        int candidate = nums[0];
        int counter = 1;
        
        for(int i =1;i<nums.length;i++){
        
            if(nums[i] == candidate){
                counter++;
            }else{
                counter--;
            }
            
            if(counter==0){
                candidate = nums[i];
                counter = 1;
            }
            
        }
        
        return candidate;

    }
}
